src/BasicAuthRequest.jl

```diff
@@ -1,32 +1,27 @@
 module BasicAuthRequest
 
 using ..Base64
-
-import ..Layer, ..request
 using URIs
-using ..Pairs: getkv, setkv
+import ..Messages: setheader, hasheader
 import ..@debug, ..DEBUG_LEVEL
 
+export basicauthlayer
 """
-    request(BasicAuthLayer, method, ::URI, headers, body) -> HTTP.Response
+    basicauthlayer(req) -> HTTP.Response
 
 Add `Authorization: Basic` header using credentials from url userinfo.
 """
-abstract type BasicAuthLayer{Next <: Layer} <: Layer{Next} end
-export BasicAuthLayer
-
-function request(::Type{BasicAuthLayer{Next}},
-                 method::String, url::URI, headers, body; kw...) where Next
-
-    userinfo = unescapeuri(url.userinfo)
-
-    if !isempty(userinfo) && getkv(headers, "Authorization", "") == ""
-        @debug 1 "Adding Authorization: Basic header."
-        setkv(headers, "Authorization", "Basic $(base64encode(userinfo))")
+function basicauthlayer(handler)
+    return function(req; basicauth::Bool=true, kw...)
```

Comment:
What does the stack traces look like now when everything are anonymous functions?

Reply:
also would it improve stacktraces if these were named functions?

e.g. 
```suggestion
    return function basicauthhandler(req; basicauth::Bool=true, kw...)
```
so they have names like `(::var"#basicauthhandler#5") (generic function with 1 method)`

Reply:
I think the backtraces are as legible as they used to be: the use of closures means we still see a nested list of the layers, and the use of a module per layer means its pretty clear where the layer is coming from.

![image](https://user-images.githubusercontent.com/2896623/159518991-ec7bb3eb-74f6-45fe-b8f8-54bad00fc813.png)